databuilder/query_engines/sqlite.py

```diff
@@ -123,14 +156,112 @@ def get_sql_select_column(self, node):
     # table references
+    @get_sql.register(SelectTable)
     @get_sql.register(SelectPatientTable)
     @cache
     def get_sql_select_table(self, node):
         return self.backend.get_table_expression(node.name)
 
+    # We ignore Filter and Sort operations completely at this point in the code and just
+    # pass the underlying table reference through. It's only later, when building the
+    # SELECT query for a given Frame, that we make use of these. This is in order to
+    # mirror the semantics of SQL whereby columns are selected directly from the
+    # underlying table and filters and sorts are handled separately using WHERE/ORDER BY
+    # clauses.
+    @get_sql.register(Sort)
+    @get_sql.register(Filter)
+    def get_sql_sort_and_filter(self, node):
+        return self.get_sql(node.source)
+
+    @get_sql.register(AggregateByPatient.Sum)
+    def get_sql_sum(self, node):
+        return self.aggregate_by_patient(node.source, sqlalchemy.func.sum)
+
+    @get_sql.register(AggregateByPatient.Exists)
+    def get_sql_exists(self, node):
+        return self.aggregate_by_patient_non_nullable(
+            node.source, sqlalchemy.literal(True), empty_value=False
+        )
+
+    @get_sql.register(AggregateByPatient.Count)
+    def get_sql_count(self, node):
+        return self.aggregate_by_patient_non_nullable(
+            node.source, sqlalchemy.func.count("*"), empty_value=0
+        )
+
+    def aggregate_by_patient(self, source_node, aggregation_func):
+        domain = get_domain(source_node)
+        query = self.get_select_query_for_domain(domain)
+        expression = self.get_sql(source_node)
+        query = query.add_columns(aggregation_func(expression).label("value"))
+        query = query.group_by(query.selected_columns[0])
+        query = prepare_query(query)
+        aggregated_table = self.reify_query(query)
+        return aggregated_table.c.value
+
+    def aggregate_by_patient_non_nullable(self, node, aggregation_func, empty_value):
+        # These aggregation functions don't take an argument as they operate over the
+        # entire query, so we wrap them in a lambda which swallows the argument
+        value = self.aggregate_by_patient(node, lambda _: aggregation_func)
+        # These aggregations are also guaranteed not to be null-valued, even when
+        # combined with data involving patient_ids for which they're not defined
+        return sqlalchemy.func.coalesce(value, empty_value)
+
+    @get_sql.register(PickOneRowPerPatient)
+    def get_sql_pick_one_row_per_patient(self, node):
+        domain = get_domain(node.source)
+        query = self.get_select_query_for_domain(domain)
+
+        # TODO: Really we only want to select the columns from the base table which
+        # we're actually going to use. In the old world we did some pre-processing of
+        # the graph to make this information available at the point we need it. But for
+        # now we just select everything for the base table.
```

Comment:
This seems like a duplicate of the TODO in the function which contains different information. Maybe merge them?